spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2002-2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.mongodb.metadata;
+
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.mongodb.core.query.Update;
+import org.springframework.integration.metadata.ConcurrentMetadataStore;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import com.mongodb.WriteResult;
+
+/**
+ * MongoDbMetadataStore implementation of {@link ConcurrentMetadataStore}.
+ * Use this {@link org.springframework.integration.metadata.MetadataStore} to
+ * achieve meta-data persistence shared across application instances and
+ * restarts.
+ *
+ * @author SenthilArumugam SamirajPanneerSelvam
+ * @since 4.0
+ *
+ */
+public class MongoDbMetadataStore implements ConcurrentMetadataStore {
+
+	private MongoTemplate template;
+	private final static String DEFAULT_COLLECTION_NAME = "metadatastore";
+	private String collectionName;
+
+	/**
+	 * 
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
+	 * @param collectionName
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		this.template = template;
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
+		this(template, null);
+	}
+
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore fileInfo = new Datastore(key, value);
+		if (!template.exists(query, collectionName)) {
+			template.save(fileInfo, collectionName);
+		}
```

Comment:
No, the `save` for `put` is enough. See `RedisMetadataStore` as an example.
Any `put` to the `MetadataStore` should store requested value for the `key`. That's all.
Thanks to the MongoDB `save` operation we have:
```
* Update an existing document or insert a document depending on the parameter. If the document does not contain an '_id' field, then
* the method performs an insert with the specified fields in the document as well as an '_id' field with a unique objectid value. If
* the document contains an '_id' field, then the method performs an upsert querying the collection on the '_id' field: <ul> <li>If a
* document does not exist with the specified '_id' value, the method performs an insert with the specified fields in the document.</li>
* <li>If a document exists with the specified '_id' value, the method performs an update, replacing all field in the existing record
* with the fields from the document.</li> </ul>
```
From here I'd prefer to make that `key` as a `_id` field:
```
* If you object has an "Id' property, it will be set with the generated Id from MongoDB. If your Id property is a
* String then MongoDB ObjectId will be used to populate that string. 
```

And yes, I think the `BasicDBObject` is the best choice for this metadata domain model. With two field: `_id` and `value`